### PR TITLE
Add session option handling to client

### PR DESCRIPTION
**Pull request by stevelorddremio**
Adds a new option, -projectId, to the list of command line options that uses DoAction SessionOptions to set project id.

### CODE REVIEW

File: java/pom.xml [hunk 1]
```diff
@@ -11,8 +11,8 @@
 
     <properties>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
-        <arrow.version>3.0.0</arrow.version>
-        <dep.guava.version>20.0</dep.guava.version>
+        <arrow.version>16.1.0-20240904212746-09241242ea-dremio</arrow.version>
```
Comment (howareyouman) [inline java/pom.xml:14]: Should we leave a public version of arrow? This is kind of an early preview one?

File: java/pom.xml [hunk 2]
```diff
@@ -73,6 +68,16 @@
 
     </dependencies>
 
+    <repositories>
+        <repository>
+            <snapshots>
+                <enabled>false</enabled>
+            </snapshots>
+            <id>dremio-public</id>
+            <url>https://repo.drem.io/content/repositories/internal-releases/</url>
```
Comment (howareyouman) [inline java/pom.xml:77]: Are you sure that we can provide links to dremio internal repositories in the public repo?

File: java/src/main/java/com/adhoc/flight/client/AdhocFlightClient.java [hunk 3]
```diff
@@ -55,21 +63,26 @@
...
 import com.adhoc.flight.utils.QueryUtils;
 import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableMap;
 
 /**
  * Adhoc Flight Client encapsulating an active FlightClient and a corresponding
  * CredentialCallOption with a bearer token for subsequent FlightRPC requests.
  */
 public final class AdhocFlightClient implements AutoCloseable {
+  public static final String PROJECT_ID_KEY = "project_id";
```
Comment (howareyouman) [inline java/src/main/java/com/adhoc/flight/client/AdhocFlightClient.java:73]: Should we expose this parameter out from the arrow flight client and not create it here?

File: java/src/main/java/com/adhoc/flight/client/AdhocFlightClient.java [hunk 4]
```diff
@@ -336,13 +363,27 @@ public void runQuery(final String query,
                        final @Nullable File fileToSaveTo,
                        final boolean printToConsole) throws Exception {
 
+    if (projectId != null) {
+      SetSessionOptionsResult res1 = client.setSessionOptions(
+          createSetSessionOption(PROJECT_ID_KEY, projectId), bearerToken, headerCallOption);
+      if (res1.hasErrors()) {
```
Comment (howareyouman) [inline java/src/main/java/com/adhoc/flight/client/AdhocFlightClient.java:369]: Should we transfer this exception out from here and not write smth in the out stream?

File: java/src/main/java/com/adhoc/flight/client/AdhocFlightClient.java [hunk 5]
```diff
@@ -336,13 +363,27 @@ public void runQuery(final String query,
...
     try (final FlightStream flightStream = getStream(flightInfo, bearerToken, headerCallOption);
          final OutputStream outputStream =
-             fileToSaveTo == null ? null : new BufferedOutputStream(new FileOutputStream(fileToSaveTo))) {
+             fileToSaveTo == null ? null : new BufferedOutputStream(
+                 Files.newOutputStream(fileToSaveTo.toPath()))) {
       writeToOutputStream(
           flightStream, allocator, outputStream, printToConsole ? QueryUtils::printResults : null);
     }
+    if (projectId != null) {
+      client.closeSession(new CloseSessionRequest(), bearerToken, headerCallOption);
```
Comment (howareyouman) [inline java/src/main/java/com/adhoc/flight/client/AdhocFlightClient.java:385]: Is it better to use try with finally? Should we close the session with an empty projectId as well?

**Review by howareyouman — CHANGES_REQUESTED** [review]
Could you add some tests to see that `projectId` is used in the client?

File: java/pom.xml (hunk 1 above)
Comment (stevelorddremio) [reply java/pom.xml:14]: This is the version that dremio is built against. It's available in the oss version, but point taken, we should be able to update to 17.0.0 release.

File: java/src/main/java/com/adhoc/flight/client/AdhocFlightClient.java (hunk 3 above)
Comment (stevelorddremio) [reply java/src/main/java/com/adhoc/flight/client/AdhocFlightClient.java:73]: No. project_id is a Dremio concept and so won't be defined in the Arrow Flight client.

File: java/src/main/java/com/adhoc/flight/client/AdhocFlightClient.java (hunk 5 above)
Comment (stevelorddremio) [reply java/src/main/java/com/adhoc/flight/client/AdhocFlightClient.java:385]: No, if setSessionOptions has errors a session will not be created.
We won't create a session unless there is an option to add. A session doesn't have meaning without options.

**Comment by stevelorddremio** [comment]
> Could you add some tests to see that `projectId` is used in the client?

This will take a l

File: java/pom.xml (hunk 2 above)
Comment (stevelorddremio) [reply java/pom.xml:77]: Not needed as I now use public repo available version.

File: java/src/main/java/com/adhoc/flight/client/AdhocFlightClient.java (hunk 4 above)
Comment (stevelorddremio) [reply java/src/main/java/com/adhoc/flight/client/AdhocFlightClient.java:369]: Converted errors into Exception message.

**Pull request closed without merging**